File: product-brain/src/product_brain/bot/webhook.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Optional

from fastapi import FastAPI, Header, HTTPException, Request

from ..adapters import get as get_adapter
from ..config import Config
from .commands import parse_brain_command
from .queue import Queue


def build_app(config: Config, queue: Optional[Queue] = None) -> FastAPI:
    app = FastAPI(title="product-brain bot")
    queue = queue or Queue(config.queue.path)
    adapter = get_adapter(config.pm_adapter, config)

    @app.get("/healthz")
    def healthz():
        return {"ok": True, "queue": queue.depth()}

    @app.post("/webhook/aha")
    async def webhook_aha(request: Request):
        body = await request.body()
        if not adapter.verify_webhook(dict(request.headers), body):
            raise HTTPException(status_code=401, detail="invalid signature")

        event = adapter.parse_webhook(body)

        if event.kind == "comment_created" and event.comment:
            author = event.comment.author
            if author not in config.bot.allowed_users:
                return {"status": "ignored", "reason": "author not allowed"}
            cmd = parse_brain_command(event.comment.body)
            if not cmd:
                return {"status": "ignored", "reason": "no /brain command"}
            if not event.ticket_id:
                return {"status": "ignored", "reason": "no ticket id"}
            queue.enqueue(
                ticket_id=event.ticket_id,
                command=cmd.command,
                trigger=f"comment:/brain {cmd.command}",
                requester=author,
                payload={"args": cmd.args},
            )
            return {"status": "queued", "command": cmd.command}

        if event.kind == "ticket_status_changed" and event.ticket_id:
            ticket = adapter.fetch_ticket(event.ticket_id)
            if config.bot.opt_in_label not in ticket.labels:
                return {"status": "ignored", "reason": "no opt-in label"}
            if config.bot.kill_switch_label in ticket.labels:
                return {"status": "ignored", "reason": "kill switch on"}
            queue.enqueue(
                ticket_id=event.ticket_id,
                command="groom",
                trigger=f"status:{event.prev_status}->{event.new_status}",
                requester="system",
                payload={"auto": True},
            )
            return {"status": "queued", "command": "groom"}

        return {"status": "ignored", "kind": event.kind}

    return app

File: product-brain/src/bot/webhook.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import Fastify, { type FastifyInstance } from "fastify";
import { getPmAdapter } from "../adapters/index.js";
import type { Config } from "../config.js";
import { parseBrainCommand } from "./commands.js";
import { Queue } from "./queue.js";

export function buildApp(config: Config, queue?: Queue): FastifyInstance {
  const app = Fastify({ logger: false });
  const q = queue ?? new Queue(config.queue.path);
  const adapter = getPmAdapter(config.pmAdapter, config);

  app.get("/healthz", () => ({ ok: true, queue: q.depth() }));

  app.addContentTypeParser(
    "application/json",
    { parseAs: "buffer" },
    (_req, body, done) => done(null, body),
  );

  app.post("/webhook/source-merge", async (request, reply) => {
    const raw = request.body as Buffer;
    const secret = process.env["PRODUCT_BRAIN_SOURCE_MERGE_SECRET"] ?? "";
    const sent = String(request.headers["x-pb-signature"] ?? "");
    if (secret) {
      const mac = createHmac("sha256", secret).update(raw).digest("hex");
      if (mac.length !== sent.length || !timingSafeEqual(Buffer.from(mac), Buffer.from(sent))) {
        return reply.code(401).send({ error: "invalid signature" });
      }
    }
    let payload: Record<string, unknown>;
    try {
      payload = JSON.parse(raw.toString("utf8")) as Record<string, unknown>;
    } catch {
      return reply.code(400).send({ error: "invalid json" });
    }
    const repoName = payload["repo"];
    const headSha = payload["head_sha"];
    const sinceSha = payload["since_sha"];
    if (!repoName || typeof repoName !== "string") {
      return reply.code(400).send({ error: "missing repo" });
    }
    try {
      config.repo(repoName);
    } catch {
      return reply.code(400).send({ error: `unknown repo: ${repoName}` });
    }
    const headPrefix = typeof headSha === "string" ? headSha.slice(0, 7) : "?";
    q.enqueue(
      `source:${repoName}`,
      "source-merge",
      `source-merge:${repoName}@${headPrefix}`,
      "ci",
      { repo: repoName, head_sha: headSha, since_sha: sinceSha },
    );
    return { status: "queued", repo: repoName };
  });

  app.post("/webhook/aha", async (request, reply) => {
    const raw = request.body as Buffer;
    const headers = request.headers as Record<string, string>;
    if (!adapter.verifyWebhook(headers, raw)) {
      return reply.code(401).send({ error: "invalid signature" });
    }
    const event = adapter.parseWebhook(raw);

    if (event.kind === "comment_created" && event.comment) {
      const author = event.comment.author;
      if (!config.bot.allowed_users.includes(author)) {
        return { status: "ignored", reason: "author not allowed" };
      }
      const cmd = parseBrainCommand(event.comment.body);
      if (!cmd) return { status: "ignored", reason: "no /brain command" };
      if (!event.ticketId) return { status: "ignored", reason: "no ticket id" };
      q.enqueue(
        event.ticketId,
        cmd.command,
        `comment:/brain ${cmd.command}`,
        author,
        { args: cmd.args },
      );
      return { status: "queued", command: cmd.command };
    }

    if (event.kind === "ticket_status_changed" && event.ticketId) {
      const ticket = await adapter.fetchTicket(event.ticketId);
      if (!ticket.labels.includes(config.bot.opt_in_label)) {
        return { status: "ignored", reason: "no opt-in label" };
      }
      if (ticket.labels.includes(config.bot.kill_switch_label)) {
        return { status: "ignored", reason: "kill switch on" };
      }
      q.enqueue(
        event.ticketId,
        "groom",
        `status:${event.prevStatus}->${event.newStatus}`,
        "system",
        { auto: true },
      );
      return { status: "queued", command: "groom" };
    }

    return { status: "ignored", kind: event.kind };
  });

  return app;
}
